pinot-connectors/presto-pinot-driver/src/main/java/org/apache/pinot/connector/presto/PinotScatterGatherQueryClient.java

```diff
@@ -0,0 +1,340 @@
+
+    // Setup QueryRouters
+    for (int i = 0; i < pinotConfig.getThreadPoolSize(); i++) {
+      _queryRouters.add(new QueryRouter(String.format("%s-%d", _prestoHostId, i), _brokerMetrics, tlsConfig));
+    }
+
+    _config = pinotConfig;
+  }
+
+  private TlsConfig getTlsConfig(Config pinotConfig) {
+    TlsConfig tlsConfig = new TlsConfig();
+    tlsConfig.setClientAuthEnabled(pinotConfig.isClientAuthEnabled());
+    tlsConfig.setTrustStoreType(pinotConfig.getTrustStoreType());
+    tlsConfig.setTrustStorePath(pinotConfig.getTrustStorePath());
+    tlsConfig.setTrustStorePassword(pinotConfig.getTrustStorePassword());
+    tlsConfig.setKeyStoreType(pinotConfig.getKeyStoreType());
+    tlsConfig.setKeyStorePath(pinotConfig.getKeyStorePath());
+    tlsConfig.setKeyStorePassword(pinotConfig.getKeyStorePassword());
+    tlsConfig.setSslProvider(pinotConfig.getSslProvider());
+    return tlsConfig;
+  }
+
+  private static <T> T doWithRetries(int retries, Function<Integer, T> caller) {
+    PinotException firstError = null;
+    for (int i = 0; i < retries; i++) {
+      try {
+        return caller.apply(i);
+      } catch (PinotException e) {
+        if (firstError == null) {
+          firstError = e;
+        }
+        if (!e.getErrorCode().isRetriable()) {
+          throw e;
+        }
+      }
+    }
+    throw firstError;
+  }
+
+  private String getDefaultPrestoId() {
+    String defaultBrokerId;
+    try {
+      defaultBrokerId = PRESTO_HOST_PREFIX + InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException e) {
+      defaultBrokerId = PRESTO_HOST_PREFIX;
+    }
+    return defaultBrokerId;
+  }
+
+  public Map<ServerInstance, DataTable> queryPinotServerForDataTable(
+      String pql,
+      String serverHost,
+      List<String> segments,
+      long connectionTimeoutInMillis,
+      boolean ignoreEmptyResponses,
+      int pinotRetryCount) {
+    BrokerRequest brokerRequest;
+    try {
+      brokerRequest = REQUEST_COMPILER.compileToBrokerRequest(pql);
+    } catch (Pql2CompilationException e) {
+      throw new PinotException(ErrorCode.PINOT_INVALID_PQL_GENERATED,
+          String.format("Parsing error with on %s, Error = %s", serverHost, e.getMessage()), e);
+    }
+
+    Map<org.apache.pinot.core.transport.ServerInstance, List<String>> routingTable = new HashMap<>();
+    routingTable.put(new org.apache.pinot.core.transport.ServerInstance(new InstanceConfig(serverHost)),
+        new ArrayList<>(segments));
+
+    // Unfortunately the retries will all hit the same server because the routing decision has already been made by
+      // the pinot broker
+    Map<ServerInstance, DataTable> serverResponseMap = doWithRetries(pinotRetryCount, (requestId) -> {
+      String rawTableName = TableNameBuilder.extractRawTableName(brokerRequest.getQuerySource().getTableName());
+      if (!_concurrentQueriesCountMap.containsKey(serverHost)) {
+        _concurrentQueriesCountMap.put(serverHost, new AtomicInteger(0));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **267 and 268**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).